cluster-autoscaler/cloudprovider/aws/api/instance_info.go

```diff
@@ -0,0 +1,356 @@
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+)
+
+const (
+	instanceInfoCacheMaxAge      = time.Hour * 6
+	awsPricingAPIURLTemplate     = "https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/%s/index.json"
+	instanceOperatingSystemLinux = "Linux"
+	instanceTenancyShared        = "Shared"
+)
+
+// InstanceInfo holds AWS EC2 instance information
+type InstanceInfo struct {
+	// InstanceType of the described instance
+	InstanceType string
+	// OnDemandPrice in USD of the ec2 instance
+	OnDemandPrice float64
+	// VCPU count of this instance
+	VCPU int64
+	// MemoryMb size in megabytes of this instance
+	MemoryMb int64
+	// GPU count of this instance
+	GPU int64
+}
+
+type httpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// NewEC2InstanceInfoService is the constructor of instanceInfoService which is a wrapper for AWS Pricing API.
+func NewEC2InstanceInfoService(client httpClient) *instanceInfoService {
+	return &instanceInfoService{
+		client: client,
+		cache:  make(instanceInfoCache),
+	}
+}
+
+type instanceInfoService struct {
+	client httpClient
+	cache  instanceInfoCache
+	sync.RWMutex
+}
+
+// DescribeInstanceInfo returns the corresponding aws instance info by given instance type and availability zone.
+func (s *instanceInfoService) DescribeInstanceInfo(instanceType string, availabilityZone string) (*InstanceInfo, error) {
+	if s.shouldSync(availabilityZone) {
+		if err := s.sync(availabilityZone); err != nil {
+			// TODO <mrcrgl> may this be tolerated for resilience
+			return nil, fmt.Errorf("failed to sync aws product and price information: %v", err)
+		}
+	}
+
+	if bucket, found := s.cache[availabilityZone]; found {
+		for _, info := range bucket.info {
+			if info.InstanceType == instanceType {
+				return &info, nil
+			}
+		}
+	}
+	return nil, fmt.Errorf("instance info not available for instance type %s in zone %s", instanceType, availabilityZone)
+}
+
+func (s *instanceInfoService) shouldSync(availabilityZone string) bool {
+	bucket, found := s.cache[availabilityZone]
+	if !found {
+		return true
+	}
+
+	return bucket.LastSync().Before(time.Now().Truncate(instanceInfoCacheMaxAge))
+}
+
+func (s *instanceInfoService) sync(availabilityZone string) error {
+	s.Lock()
+	defer s.Unlock()
+
+	start := time.Now()
+
+	bucket, found := s.cache[availabilityZone]
+	if !found {
+		bucket = new(regionalInstanceInfoBucket)
+		s.cache[availabilityZone] = bucket
+	}
+
+	response, err := s.fetch(availabilityZone, bucket.ETag)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		glog.V(4).Infof("Synchronized aws ec2 instance information for availability zone %s - took %s", availabilityZone, time.Now().Sub(start).String())
+	}()
+
+	if response == nil {
+		bucket.SetLastSync()
+		return nil
+	}
+
+	instances := make([]InstanceInfo, 0)
+	now := time.Now()
+
+	for _, product := range response.Products {
+		sku := product.SKU
```

Comment:
i would set these two vars after the filtering so that you dont have to allocate them if you `continue` the loop.